README.md
miner
=====

Localhost tunelling service wrappers

CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/bitovi/miner/tree/HEAD)

[Full Changelog](https://github.com/bitovi/miner/compare/v1.0.1...HEAD)

**Merged pull requests:**

- Package updates [\#12](https://github.com/bitovi/miner/pull/12) ([mikemitchel](https://github.com/mikemitchel))
- Package updates [\#11](https://github.com/bitovi/miner/pull/11) ([mikemitchel](https://github.com/mikemitchel))
- Add Greenkeeper badge 🌴 [\#4](https://github.com/bitovi/miner/pull/4) ([greenkeeper[bot]](https://github.com/apps/greenkeeper))

## [v1.0.1](https://github.com/bitovi/miner/tree/v1.0.1) (2022-04-05)

[Full Changelog](https://github.com/bitovi/miner/compare/v1.0.0...v1.0.1)

## [v1.0.0](https://github.com/bitovi/miner/tree/v1.0.0) (2019-04-24)

[Full Changelog](https://github.com/bitovi/miner/compare/0.2.1...v1.0.0)

**Implemented enhancements:**

- Use local interface ip addresses as hostnames [\#1](https://github.com/bitovi/miner/issues/1)

**Merged pull requests:**

- Update release process and add Changelog [\#3](https://github.com/bitovi/miner/pull/3) ([daffl](https://github.com/daffl))
- upgrade dependencies [\#2](https://github.com/bitovi/miner/pull/2) ([bmomberger-bitovi](https://github.com/bmomberger-bitovi))

## [0.2.1](https://github.com/bitovi/miner/tree/0.2.1) (2014-11-05)

[Full Changelog](https://github.com/bitovi/miner/compare/0.2.0...0.2.1)

## [0.2.0](https://github.com/bitovi/miner/tree/0.2.0) (2014-04-29)

[Full Changelog](https://github.com/bitovi/miner/compare/0.1.0...0.2.0)

## [0.1.0](https://github.com/bitovi/miner/tree/0.1.0) (2012-11-15)

[Full Changelog](https://github.com/bitovi/miner/compare/5d35bee2e485cd22591038f19a46311429e7af10...0.1.0)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


test.js
var test = require('./lib/services/pagekite');

test({ name : 'daff' }, function(err, url, process) {
	console.log('Pagekite started, go to ' + url);
	setTimeout(function() {
		console.log('killing process');
		process.kill();
	}, 10000);
});


test/test.js
var assert = require('assert');
var connect = require('connect');
var request = require('request');
var os = require('os');
var BrowserStack = require('browserstack');

var miner = require('../lib/miner.js');

describe("Miner test suite", function () {
  var isTravis = process.env.TRAVIS === 'true';
  var responseContent = 'Echo!\n';
  var port = 1337;
  var server;

  before(function (done) {
    server = connect().use(function (req, res) {
      res.writeHead(200, { 'Content-Type': 'text/plain' });
      res.end(responseContent);
    }).listen(port).on('listening', done);
  });

  after(function (done) {
    server.close(done);
  });

  describe('Default', function () {
    it('Creates a default URL', function (done) {
      miner.local({}, function (err, url) {
        assert.equal(url, 'http://localhost');
        done();
      });
    });

    it('Uses custom port and reaches host', function (done) {
      miner.local({ port: port }, function (error, url) {
        assert.ok(!error);
        assert.equal(url, 'http://localhost:' + port);
        request(url, function (error, response, body) {
          assert.ok(!error);
          assert.equal(body, responseContent);
          done();
        });
      });
    });

    it('Uses OS hostname', function (done) {
      miner.local({
        port: port,
        useOsHostname: true
      }, function (error, url) {
        assert.ok(!error);
        assert.equal(url, 'http://' + os.hostname() + ':' + port);

        // Travis won't let us access os.hostname(), only localhost
        if (isTravis) {
          return done();
        }

        request(url, function (error, response, body) {
          assert.ok(!error);
          assert.equal(body, responseContent);
          done();
        });
      });
    });
  });

  describe('Localtunnel', function () {
    it('Opens localtunnel, returns URL', function (done) {
      miner.localtunnel({
        port: port
      }, function (error, url, tunnel) {
        assert.ok(!error);
        assert.ok(url.indexOf('localtunnel.me') !== -1);
        request(url, function (error, response, body) {
          assert.ok(!error);
          assert.equal(body, responseContent);
          tunnel.kill();
          done();
        });
      });
    });
  });

  describe.skip('Browserstack', function () {
    it('Starts Browserstack tunnel, VM makes request to local server', function (done) {
      var otherPort = 1339;
      var client = BrowserStack.createClient({
        username: process.env.BROWSERSTACK_USER,
        password: process.env.BROWSERSTACK_PASSWORD
      });
      var tunnel = null;

      connect().use(function (req, res) {
        res.writeHead(200, { 'Content-Type': 'text/plain' });
        res.end('Browserstack Server\n');
        tunnel.kill();
        done();
      }).listen(otherPort).on('listening', function () {
        miner.browserstack({
          port: otherPort,
          key: process.env.BROWSERSTACK_KEY
        }, function (error, url, browserstackTunnel) {
          tunnel = browserstackTunnel;
          assert.ok(!error);
          client.createWorker({
            os: 'win',
            browser: 'ie',
            version: '10.0',
            url: url
          }, function (error) {
            assert.ok(!error);
          });
        });
      });
    });
  });

  // TODO figure out how to set up Pagekite in Travis CI
  describe.skip('Pagekite', function () {
    it('Opens pagekite, returns URL', function (done) {
      miner.pagekite({
        name: 'miner',
        port: port
      }, function (error, url, process) {
        assert.ok(!error);
        assert.ok(url.indexOf('pagekite.me') !== -1);
        request(url, function (error, response, body) {
          assert.ok(!error);
          assert.equal(body, responseContent);
          process.kill();
          done();
        });
      });
    });
  });
});


test/server.js
var http = require('http');

http.createServer(function (req, res) {
	res.writeHead(200, {'Content-Type' : 'text/plain'});
	res.end('Echo\n');
}).listen(1337, '127.0.0.1');

lib/services/pagekite.js
// https://pagekite.net/
var spawn = require('child_process').spawn;
var winston = require('winston');

module.exports = function(config, callback) {
	var url = config.domain || config.name + '.pagekite.me';
	var pagekite = spawn(config.executable || 'pagekite.py', [ config.port || 80,  url]);
	var buffer = '';
	var cb = function() {
		if(callback) {
			callback.apply(this, arguments);
			callback = null;
		}
	};
	var timeout = setTimeout(function() {
		pagekite.kill('SIGHUP');
		cb(new Error('Process timed out'));
	}, config.timeout || 30000);

	pagekite.on('exit', function(code) {
		if(code !== 0) {
			cb(new Error('Pagkekite process exited with ' + code + '\n'));
		}
	});

	pagekite.stdout.on('data', function(data) {
		buffer += data.toString();
		// winston.debug('PAGEKITE: ' + data.toString());
		if(buffer.match(new RegExp('bid=http:' + url)) !== null) {
			clearTimeout(timeout);
			cb(null, 'http://' + url, pagekite);
		}
	});

	pagekite.stderr.on('data', function(data) {
		winston.error('PAGEKITE ERROR: ' + data.toString());
	});
}


lib/services/showoff.js
// https://pagekite.net/
var spawn = require('child_process').spawn,
	winston = require('winston');

module.exports = function(config, callback) {
	var url = config.domain || config.name + '.pagekite.me';
	var pagekite = spawn(config.executable || 'pagekite.py', [ config.port || '80',  url]);
	var buffer = '';
	var cb = function() {
		if(callback) {
			callback.apply(this, arguments);
			callback = null;
		}
	};
	var timeout = setTimeout(function() {
		pagekite.kill('SIGHUP');
		cb(new Error('Process timed out'));
	}, config.timeout || 30000);

	pagekite.on('exit', function(code) {
		if(code !== 0) {
			cb(new Error('Pagkekite process exited with ' + code + '\n'));
		}
	});

	pagekite.stdout.on('data', function(data) {
		buffer += data.toString();
		// winston.debug('PAGEKITE: ' + data.toString());
		if(buffer.match(new RegExp('bid=http:' + url)) !== null) {
			clearTimeout(timeout);
			cb(null, 'http://' + url, pagekite, pagekite);
		}
	});

	pagekite.stderr.on('data', function(data) {
		winston.error('PAGEKITE ERROR: ' + data.toString());
	});
}
